kahypar/datastructure/graph.h

```diff
@@ -25,826 +25,729 @@
-    }
-    
-    size_t numEdges() const {
-        return _edges.size();
-    }
-    
-    size_t degree(const NodeID node) const  {
-        ASSERT(node < numNodes(), "NodeID " << node << " doesn't exist!");
-        return static_cast<size_t>(_adj_array[node+1]-_adj_array[node]);
-    }
-    
-    EdgeWeight weightedDegree(const NodeID node) const {
-        ASSERT(node < numNodes(), "NodeID " << node << " doesn't exist!");
-        return _weightedDegree[node];
-    }
-    
-    EdgeWeight selfloopWeight(const NodeID node) const {
-        ASSERT(node < numNodes(), "NodeID " << node << " doesn't exist!");
-        return _selfloop[node];
+    reference operator* () const {
+      return _i;
     }
-    
-    
-    EdgeWeight totalWeight() const {
-        return _total_weight;
+    NodeIDIterator& operator++ () {
+      ++_i;
+      return *this;
     }
-    
-    void setHypernodeClusterID(const HypernodeID hn, const ClusterID c_id) {
-        ASSERT(_hypernodeMapping[hn] != INVALID_NODE, "Hypernode " << hn << " isn't part of the current graph!");
-        _cluster_id[_hypernodeMapping[hn]] = c_id;
+    NodeIDIterator operator++ (int) {
+      NodeIDIterator copy(*this);
+      ++_i;
+      return copy;
     }
-    
-    void setHyperedgeClusterID(const HyperedgeID he, const ClusterID c_id, const size_t N) {
-        ASSERT(_hypernodeMapping[N+he] != INVALID_NODE, "Hyperedge " << he << " isn't part of the current graph!");
-        _cluster_id[_hypernodeMapping[N+he]] = c_id;
+
+    bool operator== (const NodeIDIterator& other) const { return _i == other._i; }
+    bool operator!= (const NodeIDIterator& other) const { return _i != other._i; }
+
+ private:
+    NodeID _i;
+  };
+
+ public:
+  static constexpr long double kEpsilon = 1e-5;
+  using NodeIterator = std::vector<NodeID>::const_iterator;
+  using EdgeIterator = std::vector<Edge>::const_iterator;
+  using IncidentClusterWeightIterator = std::vector<IncidentClusterWeight>::const_iterator;
+
+  Graph(const Hypergraph& hypergraph, const Configuration& config) :
+    _num_nodes(hypergraph.currentNumNodes() +
+               (config.preprocessing.louvain_community_detection.use_bipartite_graph ?
+                hypergraph.currentNumEdges() : 0)),
+    _num_communities(_num_nodes),
+    _total_weight(0.0L),
+    _adj_array(_num_nodes + 1),
+    _shuffle_nodes(_num_nodes),
+    _edges(),
+    _selfloop_weight(_num_nodes, 0.0L),
+    _weighted_degree(_num_nodes, 0.0L),
+    _cluster_id(_num_nodes),
+    _cluster_size(_num_nodes, 1),
+    _incident_cluster_weight(_num_nodes, IncidentClusterWeight(0, 0.0L)),
+    _incident_cluster_weight_position(_num_nodes),
+    _hypernode_mapping(hypergraph.initialNumNodes() + hypergraph.initialNumEdges(), kInvalidNode) {
+    std::iota(_shuffle_nodes.begin(), _shuffle_nodes.end(), 0);
+    std::iota(_cluster_id.begin(), _cluster_id.end(), 0);
+    if (config.preprocessing.louvain_community_detection.use_bipartite_graph) {
```

Comment:
I left that parameter there intentionally, since we want to test the heuristic that does not introduce HE-nodes for hyperedges of size 2. Then we could distinguish both versions using this parameter.